### PR TITLE
[Spec] Annotation options

### PR DESCRIPTION
**Pull request by abreslav**

### CODE REVIEW

File: spec-docs/annotation-options.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+## Inherited and Documented
+
+These two options have ratehr unclear value, and we do not supprt them in Kotlin. One can use platform-specific annotations to express them.
+
+## Appendix. Definition of kotlin.annotation
+
+``` kotlin
+package kotlin
+
+target(CLASSIFIER)
```
Comment (bashor) [inline spec-docs/annotation-options.md]: Should it be `ANNOTATION_CLASS`?

Comment (abreslav) [inline spec-docs/annotation-options.md]: Well, it's not an annotation class until this annotation is applied to it... But maybe you are right, not sure

File: spec-docs/annotation-options.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
+# Annotation Options
+
+Goals:
+* Support annotation options, such as retention policy and trageting
```
Comment (VladimirReshetnikov) [inline spec-docs/annotation-options.md]: Typo: trageting -> targeting

File: spec-docs/annotation-options.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+Putting an annotation on an element that is not allowed by the specified target is a compile-time error.
+No targets specified means that all targets are accepted.
+
+> NOTE: Java has the following [targets](https://docs.oracle.com/javase/8/docs/api/java/lang/annotation/ElementType.html):
+
+
+**TODO** Open question: what about traits/classes/objects?  
+**TODO** local variables are just like properties, but local  
+
+> Possbile platform-specific targets
```
Comment (VladimirReshetnikov) [inline spec-docs/annotation-options.md]: Typo: Possbile -> Possible

File: spec-docs/annotation-options.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+* SINGLETON_FIELD for objects
+* PROPERTY_FIELD
+* (?) DEFAULT_FUNCTION
+* (?) LAMBDA_METHOD
+* PACKAGE_FACADE
+* PACKAGE_PART
+
+### Mapping onto Java
+
+Kotlin has more possible targets than Java, so there's an issue of mapping back and forth. The table above give a correspondence.
```
Comment (VladimirReshetnikov) [inline spec-docs/annotation-options.md]: give -> gives

File: spec-docs/annotation-options.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+### Syntax
+
+It makes sense to use `kotlin.target` explicitly in Kotlin code:
+
+``` kotlin
+target(EXPRESSION, TYPE)
+annotation class MyAnn
+```
+
+> An alternative would be to make target a aproperty of `kotlin.annotation`, but then we'd
```
Comment (VladimirReshetnikov) [inline spec-docs/annotation-options.md]: aproperty -> property

File: spec-docs/annotation-options.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+* be non-uniform with Java, thus making it harder to figure how to make a Java annotation Kotlin-friendly
+
+## Retention
+
+> NOTE: Retention is a Java-specific concern, more or less. CLR retains all attributes at runtime, and JS too
+
+It makes a lot of sense to make `RUNTIME` the default retention.
+
+Since `RetentionPolicy.CLASS` is not a good fit for Kotlin that has functions outside any class, it's better to have `BINARY` instead. Also, we could have use `java.lang.annotation.RetentionPolicy` anyways, since it's platform-specific. Thus, we need to have our own enum:
+
```
Comment (VladimirReshetnikov) [inline spec-docs/annotation-options.md]: could have use -> could not use (?)

File: spec-docs/annotation-options.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+| FIELD | \<same>
+| LOCAL_VARIABLE | \<same> |
+| VALUE_PARAMETER | PARAMETER |
+| CONSTRUCTOR | \<same> |
+| FUNCITON | METHOD |
+| PROPERTY_GETTER | METHOD |
+| PROPERTY_SETTER | METHOD |
+| TYPE | TYPE_USE |
+| EXPRESSION | \<no analog> |
+
```
Comment (bashor) [inline spec-docs/annotation-options.md]: What about a file as target?

File: spec-docs/annotation-options.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+| TYPE_PARAMETER | \<same>
+| PROPERTY | \<no analog> |
+| FIELD | \<same>
+| LOCAL_VARIABLE | \<same> |
+| VALUE_PARAMETER | PARAMETER |
+| CONSTRUCTOR | \<same> |
+| FUNCITON | METHOD |
+| PROPERTY_GETTER | METHOD |
+| PROPERTY_SETTER | METHOD |
+| TYPE | TYPE_USE |
```
Comment (bashor) [inline spec-docs/annotation-options.md]: Is it ok that it called the same as other Java constant? Can it mislead?

Comment (abreslav) [inline spec-docs/annotation-options.md]: I think, it's OK

File: spec-docs/annotation-options.md (hunk 7 above)
Comment (abreslav) [inline spec-docs/annotation-options.md]: Missed that, thanks!

**Comment by abreslav** [comment]
Moved to https://gist.github.com/abreslav/6ae8195ebae37ab5219351eb576e0a00

**Pull request closed without merging**